airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/api.py

```diff
@@ -0,0 +1,167 @@
+"""
+MIT License
+
+Copyright (c) 2020 Airbyte
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+"""
+
+from abc import ABC
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional
+
+import pendulum
+import requests
+from base_python import HttpStream
+
+
+class ZendeskChatStream(HttpStream, ABC):
+    url_base = "https://www.zopim.com/api/v2/"
+
+    data_field = None
+    pagination_support = False
+    limit = 100
+
+    def backoff_time(self, response: requests.Response) -> Optional[float]:
+        delay_time = response.headers.get("Retry-After")
+        if delay_time:
+            return int(delay_time)
+
+    def path(self, **kwargs) -> str:
+        return self.name
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        if self.pagination_support:
+            stream_data = self.get_stream_data(response.json())
+            if len(stream_data) == self.limit:
+                last_object_id = stream_data[-1]["id"]
+                return {"since_id": last_object_id}
+
+    def request_params(self, next_page_token: Mapping[str, Any] = None, **kwargs) -> MutableMapping[str, Any]:
+        params = {"limit": self.limit}
+        if next_page_token:
+            params.update(next_page_token)
+
+        return params
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        response_data = response.json()
+        stream_data = self.get_stream_data(response_data)
+
+        # Getting rid of duplicates
+        if self.pagination_support and response.url.find("since_id") > -1:
```

Comment:
huh, very interesting problem here. I see why we're doing this. FWIW we can always emit duplicates because Airbyte is at-least-once-delivery. But this is OK 👍🏼